### DISCUSSION TITLE
Checking out the worktree of https://github.com/rust-lang/crates.io-index is way slower than `git` on Windows

### DISCUSSION THREAD

**Discussion by NobodyXu** [GitoxideLabs/gitoxide#931] [Q&A] [answered]
In our [CI](https://github.com/cargo-bins/cargo-binstall/actions/runs/5565131387/jobs/10165245781?pr=1204#step:11:3277), cloning https://github.com/rust-lang/crates.io-index takes roughly 9 minutes:

<details>

```
 INFO Cloning::receiving pack: Enumerating objects → 146.3k objects
 INFO Cloning::receiving pack: Compressing objects → 3.3k/55.2k objects [6%]
 INFO Cloning::receiving pack: Compressing objects → 13.2k/55.2k objects [24%]
 INFO Cloning::receiving pack: Compressing objects → 25.4k/55.2k objects [46%]
 INFO Cloning::receiving pack: Compressing objects → 32.0k/55.2k objects [58%]
 INFO Cloning::receiving pack: Compressing objects → 55.2k/55.2k objects [100%]
 INFO Cloning::receiving pack::read pack → 7.3MB
 INFO Cloning::receiving pack::read pack → 15.9MB
 INFO Cloning::receiving pack::read pack → 26.6MB
 INFO Cloning::receiving pack::read pack → 37.8MB
 INFO Cloning::receiving pack::read pack → 53.0MB
 INFO Cloning::receiving pack::read pack → 63.0MB
 INFO Cloning::receiving pack::create index file::decompressing → 427.5MB
 INFO Cloning::receiving pack::create index file::decompressing → 464.6MB
 INFO ℹCloning::receiving pack::read pack → done 83.7MB in 7.21s (11.6MB/s)
 INFO ℹCloning::receiving pack::create index file::indexing → done 146.3k objects in 4.01s (36.5k objects/s)
 INFO ℹCloning::receiving pack::create index file::decompressing → done 464.6MB in 4.01s (115.8MB/s)
 INFO ℹCloning::receiving pack::create index file::Resolving → done 0.0 objects in 1.96s (0.0 objects/s)
 INFO ℹCloning::receiving pack::create index file::Decoding → done 0B in 1.96s (0B/s)
 INFO Cloning::receiving pack::create index file → 3 of 4 steps
 INFO Cloning::receiving pack::create index file → 4 of 4 steps
 INFO ✓Cloning::receiving pack::create index file::writing index file → done 4.1MB in 0.01s (349.3MB/s)
 INFO ✓Cloning::receiving pack::create index file → done 146.3k objects in 6.01s (24.3k objects/s)
 INFO Cloning::receiving pack::writing → 6.1MB
 INFO Cloning::receiving pack::writing → 13.0MB
 INFO Cloning::receiving pack::writing → 19.2MB
 INFO Cloning::receiving pack::writing → 35.1MB
 INFO Cloning::receiving pack::writing → 44.5MB
 INFO Cloning::receiving pack::writing → 50.7MB
 INFO Cloning::receiving pack::writing → 58.3MB
 INFO Cloning::receiving pack::writing → 65.1MB
 INFO Cloning::receiving pack::writing → 71.8MB
 INFO Cloning::receiving pack::writing → 83.0MB
 INFO Cloning::receiving pack::writing → 93.2MB
 INFO Cloning::receiving pack::writing → 127.4MB
 INFO Cloning::receiving pack::writing → 134.9MB
 INFO Cloning::receiving pack::writing → 139.8MB
 INFO Cloning::receiving pack::writing → 146.7MB
 INFO Cloning::receiving pack::writing → 152.0MB
 INFO Cloning::receiving pack::writing → 157.8MB
 INFO Cloning::receiving pack::writing → 162.7MB
 INFO Cloning::receiving pack::writing → 174.2MB
 INFO Cloning::receiving pack::writing → 185.9MB
 INFO Cloning::receiving pack::writing → 191.3MB
 INFO Cloning::receiving pack::writing → 207.1MB
 INFO Cloning::receiving pack::writing → 219.1MB
 INFO Cloning::receiving pack::writing → 226.3MB
 INFO Cloning::receiving pack::writing → 233.9MB
 INFO Cloning::receiving pack::writing → 244.3MB
 INFO Cloning::receiving pack::writing → 255.6MB
 INFO Cloning::receiving pack::writing → 261.5MB
 INFO Cloning::receiving pack::writing → 267.5MB
 INFO Cloning::receiving pack::writing → 274.1MB
 INFO Cloning::receiving pack::writing → 279.4MB
 INFO Cloning::receiving pack::writing → 285.9MB
 INFO Cloning::receiving pack::writing → 290.9MB
 INFO Cloning::receiving pack::writing → 300.4MB
 INFO Cloning::receiving pack::writing → 305.4MB
 INFO Cloning::receiving pack::writing → 315.0MB
 INFO Cloning::receiving pack::writing → 320.4MB
 INFO Cloning::receiving pack::writing → 330.0MB
 INFO Cloning::receiving pack::writing → 338.9MB
 INFO Cloning::receiving pack::writing → 344.4MB
 INFO Cloning::receiving pack::writing → 356.2MB
 INFO Cloning::receiving pack::writing → 376.7MB
 INFO Cloning::receiving pack::writing → 392.0MB
 INFO Cloning::receiving pack::writing → 400.2MB
 INFO Cloning::receiving pack::writing → 405.1MB
 INFO Cloning::receiving pack::writing → 417.6MB
 INFO Cloning::receiving pack::writing → 425.2MB
 INFO Cloning::receiving pack::writing → 444.2MB
 INFO Cloning::receiving pack::writing → 450.8MB
 INFO Cloning::receiving pack::writing → 457.3MB
 INFO Cloning::receiving pack::writing → 462.8MB
 INFO Cloning::receiving pack::writing → 470.1MB
 INFO Cloning::receiving pack::writing → 476.6MB
 INFO Cloning::receiving pack::writing → 484.0MB
 INFO Cloning::receiving pack::writing → 507.6MB
 INFO Cloning::receiving pack::writing → 518.4MB
 INFO Cloning::receiving pack::writing → 525.9MB
 INFO Cloning::receiving pack::writing → 531.0MB
 INFO Cloning::receiving pack::writing → 535.8MB
 INFO Cloning::receiving pack::writing → 541.9MB
 INFO Cloning::receiving pack::writing → 548.9MB
 INFO Cloning::receiving pack::writing → 554.7MB
 INFO Cloning::receiving pack::writing → 562.9MB
 INFO Cloning::receiving pack::writing → 568.6MB
 INFO Cloning::receiving pack::writing → 575.9MB
 INFO Cloning::receiving pack::writing → 583.0MB
 INFO Cloning::receiving pack::writing → 590.6MB
 INFO Cloning::receiving pack::writing → 607.0MB
 INFO Cloning::receiving pack::writing → 616.1MB
 INFO Cloning::receiving pack::writing → 623.5MB
 INFO Cloning::receiving pack::writing → 633.5MB
 INFO Cloning::receiving pack::writing → 643.0MB
 INFO Cloning::receiving pack::writing → 648.6MB
 INFO Cloning::receiving pack::writing → 656.8MB
 INFO Cloning::receiving pack::writing → 665.2MB
 INFO Cloning::receiving pack::writing → 672.1MB
 INFO Cloning::receiving pack::writing → 679.4MB
 INFO Cloning::receiving pack::writing → 686.0MB
 INFO Cloning::receiving pack::writing → 705.9MB
 INFO Cloning::receiving pack::writing → 712.9MB
 INFO Cloning::receiving pack::writing → 719.8MB
 INFO Cloning::receiving pack::writing → 727.1MB
 INFO Cloning::receiving pack::writing → 733.1MB
 INFO Cloning::receiving pack::writing → 739.2MB
 INFO Cloning::receiving pack::writing → 748.2MB
 INFO Cloning::receiving pack::writing → 762.0MB
 INFO Cloning::receiving pack::writing → 773.4MB
 INFO Cloning::receiving pack::writing → 783.6MB
 INFO Cloning::receiving pack::writing → 791.1MB
 INFO Cloning::receiving pack::writing → 799.3MB
 INFO Cloning::receiving pack::writing → 805.1MB
 INFO Cloning::receiving pack::writing → 810.5MB
 INFO Cloning::receiving pack::writing → 820.4MB
 INFO Cloning::receiving pack::writing → 877.0MB
 INFO Cloning::receiving pack::writing → 885.7MB
 INFO Cloning::receiving pack::writing → 892.2MB
 INFO Cloning::receiving pack::writing → 899.0MB
 INFO Cloning::receiving pack::writing → 962.5MB
 INFO Cloning::receiving pack::writing → 971.9MB
 INFO Cloning::receiving pack::writing → 978.3MB
 INFO Cloning::receiving pack::writing → 983.7MB
 INFO Cloning::receiving pack::writing → 990.7MB
 INFO Cloning::receiving pack::writing → 997.0MB
 INFO Cloning::receiving pack::writing → 1007.0MB
 INFO Cloning::receiving pack::writing → 1013.7MB
 INFO Cloning::receiving pack::writing → 1020.4MB
 INFO Cloning::receiving pack::writing → 1025.0MB
 INFO Cloning::receiving pack::writing → 1031.9MB
 INFO Cloning::receiving pack::writing → 1039.0MB
 INFO Cloning::receiving pack::writing → 1049.6MB
 INFO Cloning::receiving pack::writing → 1064.1MB
 INFO Cloning::receiving pack::writing → 1.1GB
 INFO Cloning::receiving pack::writing → 1.1GB
 INFO Cloning::receiving pack::writing → 1.1GB
 INFO ℹCloning::receiving pack::checkout → done 120.0k files in 480.06s (249.0 files/s)
 INFO ℹCloning::receiving pack::writing → done 1.1GB in 480.06s (2.3MB/s)
```

</details>

It seems to spend most of the time in `ℹCloning::receiving pack::checkout` and `ℹCloning::receiving pack::writing`.
I have no idea why it is so slow since we already uses `-O3` for all crates and enables `libz-ng` on windows manually.

This problem only happens on windows and does not occur on x86-64 Linux or MacOS.

**Comment by Byron** [comment]
Thanks a lot for reaching out.

While perusing the CI log, I noticed that the [log for object resolution](https://github.com/cargo-bins/cargo-binstall/actions/runs/5565131387/jobs/10165245781?pr=1204#step:11:3295) doesn't account for any object. This should be fixed [with the latest release of `prodash`](https://github.com/Byron/prodash/releases/tag/v25.0.1).

Further, what's slow is specifically the *checkout* portion of the clone operation, which creates one small file per crate at a rate of 249 files per second. It does so with efficiency in mind and can checkout the linux kernel in below a second on a linux machine, and the code for that is the same. Of course on windows there shouldn't be such a huge difference, especially since `git` seems to be able to do it more quickly which to me seems magical as it's really all about IOPs and not wasting any. My position here is that this might be more of a rate-limit problem on the CI machine itself. I can also attest that the other optimizations, like `-O3` and `libz-ng` definitely kick in - the resolution performance is as expected on CI and finishes in 2 seconds, which depends on all of that and SHA1 performance. Checkout really only needs `libz` performance, which seems to be a given.

**But why is `git` faster then?** Can it be that `git` does a bare checkout? The crates.io index shouldn't need a worktree as the files are extracted directly from the git tree, and `cargo-binstall` shouldn't need it either (but I don't know what it does with the index).

I am conducting a test on windows now with the latest version from `main` and will hopefully be back with the results soon.

**Reply by Byron** [reply, marked as answer]
After finally and seemingly (somewhat) successfully switching the Windows VM to `aarch64-*msvc` I could produce some numbers.

The baseline is set by `git` itself, and writes are performed to the VM disk which probably isn't the fastest, but it's not sluggish either.

```
❯ time git clone https://github.com/rust-lang/crates.io-index
Cloning into 'crates.io-index'...
remote: Enumerating objects: 212699, done.
remote: Counting objects: 100% (6062/6062), done.
remote: Compressing objects: 100% (2103/2103), done.
remote: Total 212699 (delta 3906), reused 5953 (delta 3810), pack-reused 206637
Receiving objects: 100% (212699/212699), 115.89 MiB | 11.88 MiB/s, done.
Resolving deltas: 100% (130810/130810), done.
Updating files: 100% (119990/119990), done.

real    1m13.562s
user    0m0.045s
sys     0m0.046s

gitoxide on  main is 📦 v0.28.0 took 1m13s
❯ git clone file://$PWD/crates.io-index/ crates-index-git
Cloning into 'crates-index-git'...
remote: Enumerating objects: 212699, done.
remote: Counting objects: 100% (212699/212699), done.
remote: Compressing objects: 100% (66357/66357), done.
Receiving objects: 100% (212699/212699), 115.89 MiB | 19.09 MiB/s, done.
remote: Total 212699 (delta 130810), reused 212699 (delta 130810), pack-reused 0
Resolving deltas: 100% (130810/130810), done.
Updating files: 100% (119990/119990), done.

gitoxide on  main is 📦 v0.28.0 took 1m14s
```

A local clone seems to take a little longer than a remote clone, fair enough, and overall we have 1m13s to beat. Unfortunately it's hard to see how long files take to checkout, but it's the majority of the time 'by feeling'.

Now comes `gix` (`gix v0.27.0-78-g9fbed4b97` with features `max`, the default):

```
❯ ./target/release/gix.exe clone crates.io-index crates-index-gix
 11:06:12 read pack done 121.5MB in 34.74s (3.5MB/s)
 11:06:12  indexing done 212.7k objects in 28.83s (7.4k objects/s)
 11:06:12 decompressing done 680.4MB in 28.83s (23.6MB/s)
 11:06:13     Resolving done 212.7k objects in 1.24s (171.1k objects/s)
 11:06:13      Decoding done 5.3GB in 1.24s (4.3GB/s)
 11:06:13 writing index file done 6.0MB in 0.01s (416.5MB/s)
 11:06:13  create index file done 212.7k objects in 30.13s (7.1k objects/s)
 11:08:04           checkout done 120.0k files in 110.74s (1.1k files/s)
 11:08:04            writing done 1.1GB in 110.74s (9.9MB/s)
HEAD:refs/remotes/origin/HEAD (implicit)
        5432a51ba4284676ab369a44d9aa3a426fbbc2a7 HEAD symref-target:refs/heads/master -> refs/remotes/origin/HEAD [new]
+refs/heads/*:refs/remotes/origin/*
        5432a51ba4284676ab369a44d9aa3a426fbbc2a7 refs/heads/master -> refs/remotes/origin/master [new]

gitoxide on  main is 📦 v0.28.0 took 2m27s
```
Overall, performance is much lower whenever IO is involved, 2m27s vs 1m14 for git. 

The `gix` run showed quite terrible pack transfer times, maybe something with communication over pipes that is slow? Git gets a good 19MB/s, but `gix` gets only `3.5MB/s`. Speed there should be limited by the producer, but I can't explain what I am seeing here, but that's fortunately [not a problem on CI](https://github.com/cargo-bins/cargo-binstall/actions/runs/5565131387/jobs/10165245781?pr=1204#step:11:3292).

When looking at the checkout speed, then it's definitely much slower than what `git` achieves. `gix` does multi-threading by default, which gets to 22k files on MacOS and usually more on Linux, but on windows it doesn't seem to be effective. To see if it's detrimental, I disabled multi-threading, just to find out that it's mostly the same speed at 1k files/s (instead of 1.1k)

```
❯ ./target/release/gix.exe -c checkout.workers=1 clone crates.io-index crates-index-gix-cw1
 11:25:10 read pack done 121.5MB in 73.01s (1.7MB/s)
 11:25:10  indexing done 212.7k objects in 66.69s (3.2k objects/s)
 11:25:10 decompressing done 680.4MB in 66.69s (10.2MB/s)
 11:25:11     Resolving done 212.7k objects in 1.10s (193.6k objects/s)
 11:25:11      Decoding done 5.3GB in 1.10s (4.8GB/s)
 11:25:11 writing index file done 6.0MB in 0.01s (434.5MB/s)
 11:25:11  create index file done 212.7k objects in 67.85s (3.1k objects/s)
 11:27:08           checkout done 120.0k files in 117.02s (1.0k files/s)
 11:27:08            writing done 1.1GB in 117.02s (9.4MB/s)
HEAD:refs/remotes/origin/HEAD (implicit)
        5432a51ba4284676ab369a44d9aa3a426fbbc2a7 HEAD symref-target:refs/heads/master -> refs/remotes/origin/HEAD [new]
+refs/heads/*:refs/remotes/origin/*
        5432a51ba4284676ab369a44d9aa3a426fbbc2a7 refs/heads/master -> refs/remotes/origin/master [new]

gitoxide on  main is 📦 v0.28.0 took 3m11s
```

### Hypothesis about slow checkouts

I think the reason for this is that `git for windows` is actually a version Microsoft maintains and/or sponsors with a patch-queue on top of `git`, the code I am looking at that probably barely runs on windows of compiled by itself. In order to reach similar performance, one would probably have to use the same special APIs during checkout at least.

I truly admit that optimizing for windows is probably something that would happen at a much, much later stage, after `gitoxide` is established everywhere else and Microsoft itself takes an interest to picking it up for their own tooling. And that interest, I think they should have at some point given their investment into Rust. And I truly hope that either I am paid for doing these optimizations or that they will contribute them.
 
### Is there a workaround?

Fortunately, all other operations are as fast as git or faster (pack resolution, for instance), and one would have to avoid checking out 120k files. Instead, these should be read directly from the git object database. As this uses a memory map I'd be thinking that the operations there are optimized by the kernel, so performance should mostly be comparable to unix systems.
If there are questions on how to do this, I'd be happy to help.

**Reply by NobodyXu** [reply]
Thank you very much!

Optimizing for windows is indeed hard and I heard that its I/O and fs perf is worse than Linux, so it's ok to be a bit slower on windows, but our windows CI took 9x the unix CI took, which is just a bit too much.

> I truly admit that optimizing for windows is probably something that would happen at a much, much later stage, after gitoxide is established everywhere else and Microsoft itself takes an interest to picking it up for their own tooling. And that interest, I think they should have at some point given their investment into Rust. And I truly hope that either I am paid for doing these optimizations or that they will contribute them.

I think maybe IOCP can help?
Its idea is kind of like io-uring but a bit less powerful and efficient.

I was able to [`tokio-iocp`](https://lib.rs/crates/tokio-iocp) on https://libs.rs

Also, I heard that Windows now also start to support io-uring, which is something worth considering since it would speedup check out both Linux and Windows, if that is the case.

> Fortunately, all other operations are as fast as git or faster (pack resolution, for instance), and one would have to avoid checking out 120k files. Instead, these should be read directly from the git object database. As this uses a memory map I'd be thinking that the operations there are optimized by the kernel, so performance should mostly be comparable to unix systems.

That will be appreciated!

`cargo-binstall` now support `--index` and `--registry` for alternative registry, so I definitely want to avoid checking out the git registry like https://github.com/rust-lang/crates.io-index , since the `.git` for crates.io-index is only ~50M but fully checked out repository is >1G.

If there is anyway to lazily read objects as needed, then it would be awesome, we can also use this for `--git`.

**Reply by Byron** [reply]
> Optimizing for windows is indeed hard and I heard that its I/O and fs perf is worse than Linux, so it's ok to be a bit slower on windows, but our windows CI took 9x the unix CI took, which is just a bit too much.

Indeed, this is why it's probably best to avoid cloning (with worktree) entirely on that platform or use `git` for that specifically if tests rely on it (and if possible). Windows definitely won't get faster anytime soon for `gitoxide` unless it turns out to be something small that can be done. Changing all IO to io-uring is probably an undertaking that is entirely separate, and nothing I'd do without knowing how `git` on windows gets to be this fast in the first place.

> If there is anyway to lazily read objects as needed, then it would be awesome, we can also use this for --git

If you point me to the place where the full checkout is needed, I could probably figure out the rest from there as long as a git repository is present that contains the crates-index. It's not a big deal at all (*probably not even famous last words ;)*).

**Reply by NobodyXu** [reply]
> If you point me to the place where the full checkout is needed, I could probably figure out the rest from there as long as a git repository is present that contains the crates-index. It's not a big deal at all (probably not even famous last words ;)).

Here is the [code we use for cloning and checkout](https://github.com/cargo-bins/cargo-binstall/blob/main/crates/binstalk/src/helpers/git.rs).

It is then used in [`git_registry.rs`] and [`resolve.rs`].
In both scenario, they try to open a few files (could contain directory) and read their contents into memory and parse them.

In [`git_registry.rs`], we need to read `config.json` and then read a registry file for each crate queried [here](https://github.com/cargo-bins/cargo-binstall/blob/963e9e97add0a681f117fe21c6b3fde074706ac7/crates/binstalk/src/drivers/registry/git_registry.rs#L78).

In [`resolve.rs`], we would call [`load_manifest_from_workspace`](https://github.com/cargo-bins/cargo-binstall/blob/963e9e97add0a681f117fe21c6b3fde074706ac7/crates/binstalk/src/helpers/cargo_toml_workspace.rs#L20) to load the find the crate requested in the workspace.

[`git_registry.rs`]: https://github.com/cargo-bins/cargo-binstall/blob/963e9e97add0a681f117fe21c6b3fde074706ac7/crates/binstalk/src/drivers/registry/git_registry.rs#L40
[`resolve.rs`]: https://github.com/cargo-bins/cargo-binstall/blob/963e9e97add0a681f117fe21c6b3fde074706ac7/crates/binstalk/src/ops/resolve.rs#L374

**Reply by NobodyXu** [reply]
For example, the registry path of `serde` is `se/rd/serde` and is a regular file that can be read.

**Reply by Byron** [reply]
Thanks a lot for the guidance - here is [the PR](https://github.com/cargo-bins/cargo-binstall/pull/1207).
